compiler/rustc_hir_typeck/src/fn_ctxt/suggestions.rs

```diff
@@ -3365,4 +3365,113 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
             err.span_label(block.span, "this block is missing a tail expression");
         }
     }
+
+    pub(crate) fn annotate_incorrect_or_expr(
+        &self,
+        diag: &mut Diag<'_>,
+        expr: &'tcx hir::Expr<'tcx>,
+        expr_ty: Ty<'tcx>,
+        expected_ty: Ty<'tcx>,
+    ) {
+        if expected_ty != self.tcx.types.bool {
+            return;
+        }
+        let hir::Node::Expr(&hir::Expr {
+            kind:
+                hir::ExprKind::Binary(
+                    hir::BinOp { node: hir::BinOpKind::Or, span: binop_span },
+                    lhs,
+                    rhs,
+                ),
+            hir_id: parent_hir_id,
+            span: full_span,
+            ..
+        }) = self.tcx.parent_hir_node(expr.hir_id)
+        else {
+            return;
+        };
```

Comment:
I think it would be best if we documented this code. It took some time for me to fully realize what we are doing here. Consider documenting that:

* We start to see if a suggestion is applicable when `b` in `x == a || b` isn't a `bool`. (type mismatch)
* We don't try to suggest when `b` cannot be coerced to `x`.